buildtest/config_schema.json

```diff
@@ -0,0 +1,56 @@
+{
+    "$id": "https://raw.githubusercontent.com/HPC-buildtest/buildtest-framework/devel/buildtest/config_schema.json",
+    "$schema": "http://json-schema.org/draft-07/schema#",
+    "title": "Buildtest Configuration Schema Definition",
+    "type": "object",
+    "defs": {
+        "alphanumeric": {
+            "type": "string",
+            "pattern": "([a-zA-Z0-9_]|-)+"
+        },
+        "vars": {
+            "type": "object",
+            "minItems": 1,
+            "items": {
+                "type": "object",
+                "minItems": 1,            
+                "propertyNames": {
+                    "pattern": "^[A-Za-z_][A-Za-z0-9_]*$"
+                }
+            }        
+        }
+    },
+    "properties": {
+        "editor": { 
+            "type": "string",
+            "enum": ["vi","vim","nano","emacs"]
+        },
+        "queues": {
```

Comment:
Possibly we should call this executors? Queues limits to job managers, while executors also extends to cloud "queues." For example, snakemake uses executors, and SLURM/SGE/ other standard HPC queues are supported. If you are modeling a queue as explicitly different than an executor, then I'd be curious to know how an executor would be defined. I think a cleaner solution would be to have an upper level executor that could support either. Take a look at snakemake/executor.py to see how it works. Akin to how we have a BuilderBase for script, we probably could have an ExecutorBase that a custom executor extends.

Reply:
I purposely named them as `queues` to confirm to the HPC workload managers. Most if not all except Slurm refers to queues as `partition` which is their way of calling queues. See https://slurm.schedmd.com/rosetta.pdf

I'd prefer to stick to `queue` and not introduce `executors`. Currently, cloud is out of scope of buildtest. I foresee buildtest to integrate with job-scheduler so buildtest can automate job submission with default behavior is to run test locally if no scheduler is available. 

The purpose of `queues` is to allow user to specify one or more queue including `localhost` which could be default where jobs can be run. In subsequent PR what I had in mind was to have a key to select one or more queues to run the test.  Something like 

```
version: 0.0.1
hostname:
  type: script
  queues: [ "localhost", "admin" ]
  run: "hostname"
```

Again this example is a bit too simplistic, but first one ``localhost`` which is running test locally, the ``admin`` queue is of type ``scheduler: slurm``  with a ``launcher: sbatch`` which will run same test using ``sbatch``. 

I think there needs to be **keys** inside one of the schemas (script, etc...) which allows tweaking of scheduler configuration when writing test. This requires a bit more thought.

Reply:
Fair enough! I would still like to see validation for queue specific types, because it's always going to be the case that you see srun with slurm, for example. Having validation on the level of the queue also ensures that it's easy to add "other queue-like things" (ahem executors) that you might not think are important, but not everyone agrees with you about that.

Reply:
`executors` could be something we can put on roadmap, but will need some use-case and how it's intended to work. Right now this was first attempt and will require more revisions to schema. This also addresses the fact we didn't have any check on buildtest configuration (``settings.yml``) see #158. 

On a unrelated note, I foresee no point in having ``buildtest show --config`` when ``buildtest config view`` is showing the content in json.

I intended ``buildtest show --config`` to represent data like key/value pair that can be easily interpreted by user instead of reading a json or yaml file. The output style was inspired by ``eb --show-config`` see https://easybuild.readthedocs.io/en/latest/Configuration.html#overview-of-current-configuration-show-config-show-full-config

The output of this mess looks like 

```
gitpod /workspace/buildtest-framework $ buildtest show --config
editor                                             = vim
queues                                             = [{'name': 'localhost', 'description': 'submit job locally', 'scheduler': 'local', 'launcher': 'local'}, {'name': 'admin', 'description': 'Admin Queue', 'scheduler': 'slurm', 'launcher': 'sbatch', 'options': ['#SBATCH -n 1'], 'vars': {'FOO': 'BAR'}}]
``` 

If it makes sense, i can remove this option in this PR as well.

Reply:
Sure thing, removing show config is okay with me!

I would still like to see validation for queue specific types. If we are going to implement this, it makes sense to do it right the first time.

Reply:
ok i will remove ``buildtest show --config``. 

I think i understand the validation that is missing I will get that done. The ``launcher`` need to be compatible with ``scheduler`` type so that user can't say ``launcher: bsub`` and specify ``scheduler: slurm``.

Ill get to this tomorrow hopefully.

Reply:
Awesome! I'm recording a podcast this morning followed by meetings, but I'll be around later today to review. This is pseudo code, but I am thinking something like:

```json
{
    "type": "array",
    "items": {
        "oneOf": [
            {"slurm":  "$ref": "#slurm" }, 
            {"sge": [  "$ref": "#sge" ] }
        ]
    }  
}
```
(Note sure if the ref is correct, but the gist is that you have a reference for an item under definition, and each of these items can be specific to the scheduler!  There are a bunch of examples here https://json-schema.org/understanding-json-schema/structuring.html#using-id-with-ref. This way, the properties that are decided on for within some scheduler don't limit us for future development of other schedulers, and if an institution really wanted to, they could modify the config to add a custom scheduler (for example TACC has a tool on one of their clusters called the launcher that is specific to TACC).

Gotta run!